source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+discriminated using user-defined boxcar and discrimnation ``kernel``s.
+
+.. code-block:: javascript
+
+    // Use a boxcar function to generate IQ data from raw waveform
+    kernel boxcar(waveform input) -> complex[64];
+    // Use a linear discriminator to generate bits from IQ data
+    kernel discriminate(complex[64] iq) -> bit;
+
+    defcal measure $0 -> bit {
+        // Define the channels
+        txchannel m0 = txch($0, "measure");
+        rxchannel cap0 = rxch($0, "capture");
+
+        // Force time of carrier to 0 for consistent phase for discrimination.
+        frame stimulus_frame = newframe(5e9, 0);
+        frame capture_frame = newframe(5e9, 0);
+
+        // Apply measurement stimulus
+        waveform meas_wf = gaussian_square(1.0, 16000dt, 262dt, 13952dt);
+
+        // Play the stimulus
+        play(m0, meas_wf, stimulus_frame);
+        // Align measure and capture channels
+        barrier(stimulus_frame, capture_frame);
+        // Capture transmitted data after interaction with measurement resonator
+        // kernel capture(channel chan, frame capture_frame) -> waveform;
+        waveform raw_output = receive(cap0, meas_pulse.duration);
+
+        // Kernel and discriminate
+        complex[32] iq = boxcar(raw_output);
+        bit result = discriminate(iq);
+
+        return result;
+    }
+
+
+Timing
+------
+
+Each frame maintains its own "clock". When a pulse is played the clock for
+that frame advances by the length of the pulse.
+
+For frames, everything behaves analogous to qubits in the
+`Delays <delays.html>`_ section of this specification. There are however some
+small differences.
+
+The ``delay`` instruction may take a frame instead of a qubit. The ``barrier``
+instruction may also take a list of frames instead of a list of qubits and aligns the time
+of the clocks given as arguments.
+
+``defcal`` blocks have an implicit ``barrier`` on every frame that enters the block,
+meaning that those clocks are guaranteed to be aligned at the start of the block.
+These blocks also need to have a well-defined length, similar to the ``boxas`` block.
+
+.. code-block:: javascript
+
+   cal {
+     waveform p = ...; // some 100dt waveform
+     frame driveframe1 = newframe(5.0e9, 0);
+     frame driveframe2 = newframe(6.0e9, 0);
+   }
+
+   defcal aligned_gates {
+     // driveframe1 and driveframe2 used in this defcal, so clocks are aligned
+     play(tx0, p, driveframe1);
+     delay[20dt] driveframe1;
+     // Clocks now unaligned by 120dt, so we use a `barrier` to re-align
+     barrier(driveframe1, driveframe2);
+     // `driveframe2` will now play a pulse 120dt after `driveframe1` finishes playing
+     play(tx0, p, driveframe2);
+   }
+
+Examples
+--------
+
+Cross-resonance gate
+~~~~~~~~~~~~~~~~~~~~
+
+
+.. code-block:: javascript
+
+  cal {
+     frame frame0 = newframe(5.0e9, 0);
+  }
+
+  defcal cross_resonance $0 $1 {
+      // Access globally defined channels
+      channel d0 = txch($0, "drive");
```

Comment:
```suggestion
      txchannel d0 = txch($0, "drive");
```

Reply:
Ignoring since the decision was to use channel instead of txchannel/rxchannel